File: packages/node-sdk/src/event.ts
import net from "node:net";
import { ProtocolError, TransportError } from "./errors.js";
import { EVENT_PROTOCOL, EventEnvelope, PROTOCOL_VERSION } from "./protocol.js";

export interface EventFilter {
  name?: string;
  name_prefix?: string;
  publisher_kinds?: string[];
}

const MAX_FRAME_SIZE = 64 * 1024;

export class EventSubscription {
  constructor(
    private readonly socket: net.Socket,
    private readonly reader: FrameReader,
    readonly filters: EventFilter[],
    readonly timeoutMs: number,
  ) {}

  async recv(timeoutMs?: number): Promise<EventEnvelope> {
    const value = await this.reader.next(timeoutMs ?? this.timeoutMs);
    if (value.kind !== "delivery") throw new ProtocolError(`expected event delivery, got ${String(value.kind)}`);
    return value.event as EventEnvelope;
  }

  /** Bounded receive: at most `maxEvents`, returning early on aggregate timeout. */
  async receive(maxEvents: number, timeoutMs?: number): Promise<EventEnvelope[]> {
    if (!Number.isInteger(maxEvents) || maxEvents <= 0) throw new TypeError("maxEvents must be a positive integer");
    const events: EventEnvelope[] = [];
    const start = Date.now();
    const budget = timeoutMs ?? this.timeoutMs;
    while (events.length < maxEvents) {
      const remaining = budget - (Date.now() - start);
      if (remaining <= 0) break;
      try {
        events.push(await this.recv(remaining));
      } catch (error) {
        if (error instanceof TransportError && error.message === "event recv timeout") break;
        throw error;
      }
    }
    return events;
  }

  async nextFileDrop(imagesOnly = true, timeoutMs?: number): Promise<EventEnvelope> {
    for (;;) {
      const event = await this.recv(timeoutMs);
      if (event.name !== "ui.file_drop.accepted") continue;
      const payload = event.payload as { is_image?: boolean };
      if (imagesOnly && !payload.is_image) continue;
      return event;
    }
  }

  async *typedEvents(name: string, timeoutMs?: number): AsyncGenerator<EventEnvelope> {
    for (;;) {
      try {
        const event = await this.recv(timeoutMs);
        if (event.name === name) yield event;
      } catch (error) {
        if (error instanceof TransportError && error.message === "event recv timeout") return;
        throw error;
      }
    }
  }

  close(): void { this.socket.destroy(); }
}

export class EventClient {
  constructor(readonly endpoint: string, readonly options: { origin?: string; kind?: string; instanceId?: string; timeoutMs?: number } = {}) {}

  async subscribe(filter: EventFilter): Promise<EventSubscription> {
    if (!filter.name && !filter.name_prefix) throw new TypeError("name or name_prefix is required");
    const timeoutMs = this.options.timeoutMs ?? 10000;
    const socket = await connect(this.endpoint, timeoutMs);
    const instanceId = this.options.instanceId ?? "neon3-event-client";
    await writeFrame(socket, {
      kind: "subscribe", protocol: EVENT_PROTOCOL, version: PROTOCOL_VERSION,
      request_id: `neon3-event-subscribe-${instanceId}`,
      client: { kind: this.options.kind ?? "external_host", instance_id: instanceId, pid: process.pid, origin: this.options.origin ?? "neon3-node-sdk" },
      filters: [filter], replay_from_sequence: null, max_rate_hz: null,
    });
    const reader = new FrameReader(socket);
    const ack = await reader.next(timeoutMs);
    if (ack.kind !== "ack" || ack.status !== "accepted") { socket.destroy(); throw new ProtocolError(`event subscription rejected: ${JSON.stringify(ack)}`); }
    return new EventSubscription(socket, reader, [filter], timeoutMs);
  }
}

function connect(endpoint: string, timeoutMs: number): Promise<net.Socket> {
  const match = endpoint.match(/^([^:]+):(\d+)$/);
  if (!match || (!match[1].startsWith("127.") && match[1] !== "localhost")) throw new TypeError("endpoint must be loopback host:port");
  const port = Number(match[2]);
  if (!Number.isInteger(port) || port < 1 || port > 65535) throw new TypeError("endpoint port must be between 1 and 65535");
  return new Promise((resolve, reject) => {
    const socket = net.createConnection({ host: match[1], port });
    const timer = setTimeout(() => { socket.destroy(); reject(new TransportError("event timeout")); }, timeoutMs);
    socket.once("connect", () => { clearTimeout(timer); resolve(socket); });
    socket.once("error", error => { clearTimeout(timer); reject(error); });
  });
}

function writeFrame(socket: net.Socket, value: unknown): Promise<void> {
  const payload = Buffer.from(JSON.stringify(value));
  if (payload.length > MAX_FRAME_SIZE) return Promise.reject(new TransportError("event frame_too_large"));
  return new Promise((resolve, reject) => socket.write(Buffer.concat([u32(payload.length), payload]), error => error ? reject(error) : resolve()));
}

class FrameReader {
  private buffer = Buffer.alloc(0);
  private waiters: Array<{ resolve: (value: any) => void; reject: (error: Error) => void; timer: NodeJS.Timeout | null }> = [];
  private failure: Error | undefined;

  constructor(private readonly socket: net.Socket) {
    socket.on("data", chunk => { this.buffer = Buffer.concat([this.buffer, chunk]); this.pump(); });
    socket.on("error", error => this.fail(error instanceof Error ? error : new Error(String(error))));
    socket.on("close", () => this.fail(new TransportError("event connection closed")));
  }

  next(timeoutMs?: number): Promise<any> {
    if (this.failure) return Promise.reject(this.failure);
    const value = this.take();
    if (value !== undefined) return Promise.resolve(value);
    return new Promise((resolve, reject) => {
      let timer: NodeJS.Timeout | null = null;
      if (timeoutMs !== undefined) {
        timer = setTimeout(() => {
          const index = this.waiters.findIndex(waiter => waiter.resolve === resolve);
          if (index >= 0) this.waiters.splice(index, 1);
          reject(new TransportError("event recv timeout"));
        }, timeoutMs);
      }
      this.waiters.push({ resolve, reject, timer });
    });
  }

  private take(): any | undefined {
    if (this.buffer.length < 4) return undefined;
    const size = this.buffer.readUInt32BE(0);
    if (size > MAX_FRAME_SIZE) throw new ProtocolError("event frame_too_large");
    if (this.buffer.length < size + 4) return undefined;
    const body = this.buffer.subarray(4, size + 4);
    this.buffer = this.buffer.subarray(size + 4);
    return JSON.parse(body.toString("utf8"));
  }

  private pump(): void {
    while (this.waiters.length) {
      try {
        const value = this.take();
        if (value === undefined) return;
        const waiter = this.waiters.shift()!;
        if (waiter.timer) clearTimeout(waiter.timer);
        waiter.resolve(value);
      } catch (error) { this.fail(error instanceof Error ? error : new Error(String(error))); return; }
    }
  }

  private fail(error: Error): void {
    if (this.failure) return;
    this.failure = error;
    for (const waiter of this.waiters.splice(0)) {
      if (waiter.timer) clearTimeout(waiter.timer);
      waiter.reject(error);
    }
  }
}

function u32(value: number): Buffer { const result = Buffer.allocUnsafe(4); result.writeUInt32BE(value); return result; }

File: packages/python-sdk/src/neon3_sdk/event.py
"""Persistent client for the canonical ``neon3.event`` protocol."""

from __future__ import annotations

import json
import socket
import struct
from dataclasses import dataclass
from typing import Any, Iterator

from .client import _parse_loopback_endpoint
from .errors import ProtocolError, TransportError
from .models import ClientIdentity, EventEnvelope, UiFileDropPayload

EVENT_PROTOCOL = "neon3.event"
PROTOCOL_VERSION = {"major": 1, "minor": 0}
MAX_FRAME_SIZE = 64 * 1024


@dataclass(frozen=True)
class EventFilter:
    name: str | None = None
    name_prefix: str | None = None
    publisher_kinds: tuple[str, ...] | None = None

    def to_wire(self) -> dict[str, Any]:
        return {
            "name": self.name,
            "name_prefix": self.name_prefix,
            "publisher_kinds": list(self.publisher_kinds) if self.publisher_kinds else None,
        }


class EventSubscription:
    def __init__(self, stream: socket.socket, reader: "_FrameReader", client: ClientIdentity, filters: list[EventFilter]) -> None:
        self._stream = stream
        self._reader = reader
        self.client = client
        self.filters = filters

    def recv(self) -> EventEnvelope:
        response = self._reader.read()
        if response.get("kind") != "delivery":
            raise ProtocolError(f"expected event delivery, got {response.get('kind')}")
        return EventEnvelope.from_wire(response["event"])

    def file_drops(self, *, images_only: bool = True) -> Iterator[UiFileDropPayload]:
        while True:
            event = self.recv()
            if event.name != "ui.file_drop.accepted":
                continue
            payload = UiFileDropPayload.from_wire(event.payload)
            if images_only and not payload.is_image:
                continue
            yield payload

    def close(self) -> None:
        self._stream.close()

    def __enter__(self) -> "EventSubscription":
        return self

    def __exit__(self, _type: Any, _value: Any, _traceback: Any) -> None:
        self.close()


class EventClient:
    def __init__(self, endpoint: tuple[str, int], identity: ClientIdentity, timeout_seconds: float = 10.0) -> None:
        self.endpoint = endpoint
        self.identity = identity
        self.timeout_seconds = timeout_seconds

    @classmethod
    def connect(
        cls,
        endpoint: str | tuple[str, int],
        *,
        origin: str = "neon3-python-sdk",
        kind: str = "external_host",
        instance_id: str = "neon3-event-client",
        timeout_seconds: float = 10.0,
    ) -> "EventClient":
        parsed = _parse_loopback_endpoint(endpoint)
        return cls(parsed, ClientIdentity(kind, instance_id, 0, origin), timeout_seconds)

    def subscribe(
        self,
        *,
        name: str | None = None,
        name_prefix: str | None = None,
        publisher_kinds: tuple[str, ...] | None = None,
        replay_from_sequence: int | None = None,
        max_rate_hz: int | None = None,
    ) -> EventSubscription:
        if not name and not name_prefix:
            raise ValueError("name or name_prefix is required")
        stream = socket.create_connection(self.endpoint, timeout=self.timeout_seconds)
        stream.settimeout(self.timeout_seconds)
        request_id = f"neon3-event-subscribe-{self.identity.instance_id}"
        _write_frame(stream, {
            "kind": "subscribe",
            "protocol": EVENT_PROTOCOL,
            "version": PROTOCOL_VERSION,
            "request_id": request_id,
            "client": self.identity.to_wire(),
            "filters": [EventFilter(name, name_prefix, publisher_kinds).to_wire()],
            "replay_from_sequence": replay_from_sequence,
            "max_rate_hz": max_rate_hz,
        })
        reader = _FrameReader(stream)
        ack = reader.read()
        if ack.get("kind") != "ack" or ack.get("status") != "accepted":
            stream.close()
            raise ProtocolError(f"event subscription rejected: {ack}")
        return EventSubscription(stream, reader, self.identity, [EventFilter(name, name_prefix, publisher_kinds)])


def _write_frame(stream: socket.socket, value: dict[str, Any]) -> None:
    payload = json.dumps(value, separators=(",", ":"), ensure_ascii=True).encode("utf-8")
    if len(payload) > MAX_FRAME_SIZE:
        raise TransportError("event frame_too_large")
    stream.sendall(struct.pack(">I", len(payload)) + payload)


def _read_frame(stream: socket.socket) -> dict[str, Any]:
    header = _recv_exact(stream, 4)
    size = struct.unpack(">I", header)[0]
    if size > MAX_FRAME_SIZE:
        raise ProtocolError("event frame_too_large")
    try:
        decoded = json.loads(_recv_exact(stream, size).decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        raise ProtocolError(f"invalid event JSON: {error}") from error
    if not isinstance(decoded, dict):
        raise ProtocolError("event frame must be a JSON object")
    return decoded


class _FrameReader:
    """Length-prefixed reader that preserves multiple frames in one recv."""

    def __init__(self, stream: socket.socket) -> None:
        self.stream = stream
        self.buffer = bytearray()

    def read(self) -> dict[str, Any]:
        while True:
            if len(self.buffer) >= 4:
                size = struct.unpack(">I", self.buffer[:4])[0]
                if size > MAX_FRAME_SIZE:
                    raise ProtocolError("event frame_too_large")
                if len(self.buffer) >= size + 4:
                    payload = bytes(self.buffer[4:size + 4])
                    del self.buffer[:size + 4]
                    try:
                        decoded = json.loads(payload.decode("utf-8"))
                    except (UnicodeDecodeError, json.JSONDecodeError) as error:
                        raise ProtocolError(f"invalid event JSON: {error}") from error
                    if not isinstance(decoded, dict):
                        raise ProtocolError("event frame must be a JSON object")
                    return decoded
            chunk = self.stream.recv(64 * 1024)
            if not chunk:
                raise TransportError("connection_closed")
            self.buffer.extend(chunk)


def _recv_exact(stream: socket.socket, length: int) -> bytes:
    chunks: list[bytes] = []
    remaining = length
    while remaining:
        chunk = stream.recv(remaining)
        if not chunk:
            raise TransportError("connection_closed")
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)
